### DISCUSSION TITLE
Expand AddOn to work with X3D

### DISCUSSION THREAD

**Discussion by ThomasSchumacher77** [neverhood311/Stop-motion-OBJ#143] [Ideas]
Hi Justin,
would it be possible to extend the functionality to also work with x3d file format?
Regards
Thomas

**Comment by neverhood311** [comment]
I have no experience with the x3d file format, which is why I haven't yet included it. Some questions:

- Can you describe a workflow that requires the x3d file format and couldn't be accomplished with obj, stl, or ply?
- What kinds of features does x3d support?
  - Animation?
  - Multiple objects?
  - Lights, cameras, and other non-mesh entities?
  - Scene hierarchy?
  - Actions/Tracks? (like .FBX)

I'm hesitant to add support for a file format with too many features. Since the addon would be importing many files at once, it becomes difficult to handle the complexity. For example, what would I do with a sequence of Collada files that each included animation, cameras, and lights?

If these extra features are not essential to a .x3d mesh sequence, there is a possible compromise: import each file and discard anything that isn't a mesh, material, or texture. Also discard any extra objects. This is currently how Stop Motion OBJ handles obj files, which support multiple objects.

**Reply by ThomasSchumacher77** [reply]
Hi Justin,
thanks for the quick reply.
In terms of content and capability x3d is very similar to ply. Basically contains a faceted surface with vertex color.
Blender comes with a x3d importer out of the box.
Why is it needed? x3d format is usually the format CFD users choose to get OpenFOAM (Open source CFD code, widely used) data into Blender for example. 
There are a number of blender batch scripts out there that process a series of x3d files into renders, but nothing as elegant and easy to use as your add-on.

Regards
Thomas

**Reply by neverhood311** [reply]
Ok, I'll give this a go.

**Reply by neverhood311** [reply]
Well that was easy.

I've added X3D support. Currently, it's available in a pre-release, [v2.2.0.alpha.11](https://github.com/neverhood311/Stop-motion-OBJ/releases/tag/v2.2.0.alpha.11). If you want to try it out before the final release of v2.2.0 (some time before the end of the year), then grab alpha 11. If you run into any issues, please file a bug report.

**Reply by ThomasSchumacher77** [reply]
Wow, that fast fast.
Is it possible to define (in the code) which of the objects will be read? You are currently limited to the first object in the file.
My files are usually looking like the attached.
[scene_0.63000000.zip](https://github.com/neverhood311/Stop-motion-OBJ/files/7400160/scene_0.63000000.zip)
When using the new functionality I get an empty sequence. What would be needed is the "shape" from the x3d file. The camera/view and the lights could be discarded.

Regards
Thomas

**Reply by neverhood311** [reply]
I'll take a look. In the meantime, is there a way to export the .x3d file without the lights and camera?

**Reply by ThomasSchumacher77** [reply]
Not that I found, no. The workaround would be to somehow delete those lines from hundreds of files with a script.

**Reply by ThomasSchumacher77** [reply]
I think the "problem" is, that x3d is a scene format, in contrast to ply, which is merely a mesh format. So the native Blender importer for x3d by default loads all the lights and cameras it can find.

**Reply by neverhood311** [reply]
Yeah, that's why I've been very selective about which formats to support. But I think this one could work still.

As I'm importing each file, Blender gives me a list of all the objects it just created. What if I just search through that list and take the first one that has a type of `'MESH'`? Then I'd just delete everything else. I know it's probably possible to do multi-object sequences but I just don't want to tackle that right now.
Would this be suitable for your workflow? (hoping your workflow is more-or-less a standard one)

**Reply by ThomasSchumacher77** [reply]
When I load a x3d using Blender I get the attached list of objects.
In this case my x3d included 2 separate mesh objects. Everything else is not needed, as I want to setup lights and camera from blender anyway.
[scene0_00000.0.zip](https://github.com/neverhood311/Stop-motion-OBJ/files/7400519/scene0_00000.0.zip)
![x3dimport](https://user-images.githubusercontent.com/79206222/138501223-bedfedde-e558-4637-81c1-e2fa39df8342.png)

**Reply by neverhood311** [reply]
Is it common to have .x3d files with multiple meshes like this? Or can you somehow export two separate mesh sequences from OpenFOAM?

Supporting multi-object sequences would add a fair amount of complexity to the import process. I'm not sure when I'll be able to work on that.

**Reply by ThomasSchumacher77** [reply]
Yeah,I think for now the easiest solution is to work with multiple single-mesh sequences.

**Reply by neverhood311** [reply]
Ok, give this one a try: https://github.com/neverhood311/Stop-motion-OBJ/releases/tag/v2.2.0.alpha.12

**Reply by ThomasSchumacher77** [reply]
Works! 
Thats really helpful, thanks a lot.
I hope to be able to publish something shiny soon, with this.
Regards
Thomas
